### PR TITLE
LMSR fixups and migration to Truffle 5

### PR DESCRIPTION
**Pull request by cag**
Last commit migrating to Truffle 5 can be left off or branched off if desired.

Stats diff:
```diff
@@ -28,29 +28,29 @@ Contract: PredictionMarketSystem
 
 Contract: LMSRMarketMakerFactory
   createLMSRMarketMaker:
-    min: 2617702
-    max: 2633086
-    avg: 2623206.153846154
-    med: 2617766
+    min: 2637618
+    max: 2653002
+    avg: 2642341.6923076925
+    med: 2637682
 
 Contract: LMSRMarketMaker
   fund:
-    min: 141446
-    max: 187087
-    avg: 154632.6923076923
-    med: 156446
+    min: 141424
+    max: 187065
+    avg: 154610.6923076923
+    med: 156424
   calcNetCost:
-    min: 84195
+    min: 71813
     max: undefined
     avg: NaN
     med: undefined
   trade:
-    min: 179591
-    max: 366988
-    avg: 213684.49206349207
-    med: 202285
+    min: 161146
+    max: 262185
+    avg: 177444.49206349207
+    med: 166431
   calcMarginalPrice:
-    min: 45422
+    min: 53617
     max: undefined
     avg: NaN
     med: undefined
@@ -60,7 +60,7 @@ Contract: LMSRMarketMaker
     avg: 75142
     med: 75142
   calcMarketFee:
-    min: 22618
+    min: 22596
     max: undefined
     avg: NaN
     med: undefined
@@ -79,5 +79,5 @@ Contract: WETH9
   approve:
     min: 45225
     max: 47209
-    avg: 45547.37037037037
+    avg: 45533.148148148146
     med: 45545
```

Turns out gas reduction is not as significant as originally hoped for. This uses ~20% less gas for the trade function.

calcMarginalPrice is actually _more_ expensive now, which is a bummer. The reason is that it now makes queries to pmSystem for the outcome tokens held data instead querying its local storage for its own accounting of net outcome tokens sold. Though there is theoretically only one exponentiation which has to be done, because of edge cases written in the test scenarios, what should be "one" actually has to be recalculated from the outcome tokens held data sometimes (this happens when funding is way less than the trading activity with the market maker).

Gas measurement tools don't work under the new Truffle. However, the project will now have access to solc 0.5.0.

I have also fixed up the two test cases which have been skipped for the longest time. This changeset actually addresses an issue in the original code, which had been modified with conservative rounding bounds to ensure that the trading stress tests pass. Unfortunately, the rounding error bounds also caused the market maker to price outcome tokens to be _more_ than its corresponding collateral token (by an order of 1e-18 to be fair, but still enough to cause the original tests to fail).

Also, due to the way that ERC 1155 is specified, and due to how onERC1155Received has been implemented in the MarketMaker, the attack I was concerned about with randos sending outcome tokens to the market maker to try and mess with its account should actually not be possible. However, I did disable the guard and run the stress test with random sends of outcome tokens 1000 times, and it passed nonetheless, so I believe that even if somebody were to figure out a way to send outcome tokens to the market maker without going through one of the exposed methods, it would be simply a donation to the market maker and would not affect its operations terribly much.

### CODE REVIEW

File: test/utils.js [hunk 1]
```diff
@@ -63,14 +63,13 @@ function getBlock(b) {
...
 function lmsrMarginalPrice(funding, netOutcomeTokensSold, outcomeIndex) {
-    const b = Decimal(funding.valueOf()).div(netOutcomeTokensSold.length).ln()
-
-    return Decimal(netOutcomeTokensSold[outcomeIndex].valueOf()).div(b).exp().div(
-        netOutcomeTokensSold.reduce(
-            (acc, tokensSold) => acc.add(Decimal(tokensSold.valueOf()).div(b).exp()),
-            Decimal(0)
-        )
-    ).valueOf()
+    const b = new Decimal(funding.toString()).div(Decimal.ln(netOutcomeTokensSold.length))
```
Comment (cag) [inline test/utils.js]: This liquidity parameter was calculated wrong (see the offset parenthesis)! That was the main reason those skipped tests were failing :cry:

File: test/test_market_makers.js [hunk 2]
```diff
@@ -161,8 +172,8 @@ contract('MarketMaker', function(accounts) {
                 )
             }
 
-            // Price is equal to 1
-            assert.equal(cost, tokenCount)
+            // Price is at least 1
+            assert(cost.gte(tokenCount))
```
Comment (cag) [inline test/test_market_makers.js]: This change was made because of the rounding compensation in the old code causing the market maker to price the outcome tokens more than collateral sometimes. This changeset actually permits the stricter equals check.

**Comment by InfiniteStyles** [comment]
Seems good. Going to merge this in, so that I can rebase it.
